File: scripts/day12-part1-simple.js
const fs = require("fs");

class Shape {
  constructor(coords) {
    this.coords = coords;
  }
}

function tryPlace(idx, presents, grid, width, height, shapes) {
  if (idx >= presents.length) return true;

  const shapeIdx = presents[idx];
  const shape = shapes[shapeIdx];

  for (let r = 0; r < height; r++) {
    for (let c = 0; c < width; c++) {
      let canPlace = true;
      for (const [dr, dc] of shape.coords) {
        const nr = r + dr,
          nc = c + dc;
        if (nr >= height || nc >= width || grid[nr][nc] !== -1) {
          canPlace = false;
          break;
        }
      }

      if (canPlace) {
        for (const [dr, dc] of shape.coords) {
          grid[r + dr][c + dc] = idx;
        }

        if (tryPlace(idx + 1, presents, grid, width, height, shapes)) {
          return true;
        }

        for (const [dr, dc] of shape.coords) {
          grid[r + dr][c + dc] = -1;
        }
      }
    }
  }

  return false;
}

function solve() {
  const lines = fs
    .readFileSync("../inputs/day-12-input.txt", "utf-8")
    .split("\n");
  const shapes = {};

  // Parse shapes
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    if (/^\d+:$/.test(line)) {
      const shapeIdx = parseInt(line);
      const coords = [];
      i++;
      let row = 0;

      while (
        i < lines.length &&
        lines[i].trim() &&
        !/^\d+:$/.test(lines[i]) &&
        !lines[i].includes("x")
      ) {
        for (let c = 0; c < lines[i].length; c++) {
          if (lines[i][c] === "#") {
            coords.push([row, c]);
          }
        }
        row++;
        i++;
      }
      i--;
      shapes[shapeIdx] = new Shape(coords);
    }
  }

  // Parse regions
  const regions = [];
  for (const line of lines) {
    if (line.includes("x") && line.includes(":") && /^\d/.test(line)) {
      const [dims, rest] = line.split(":");
      const [width, height] = dims.split("x").map(Number);
      const counts = rest.trim().split(/\s+/).map(Number);

      const presents = [];
      counts.forEach((count, idx) => {
        for (let j = 0; j < count; j++) {
          presents.push(idx);
        }
      });

      regions.push({ width, height, presents });
    }
  }

  let fittable = 0;

  for (const r of regions) {
    const area = r.width * r.height;
    const needed = r.presents.reduce(
      (sum, p) => sum + shapes[p].coords.length,
      0
    );

    if (needed <= area) {
      const grid = Array.from({ length: r.height }, () =>
        Array(r.width).fill(-1)
      );
      if (tryPlace(0, r.presents, grid, r.width, r.height, shapes)) {
        fittable++;
      }
    }
  }

  return fittable;
}

// Benchmark
const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solve();
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMs = (totalNanos / iterations / 1000000).toFixed(3);

console.log(`JavaScript (Day 12 Part 1 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMs} ms (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000000).toFixed(2)} s`);

File: scripts/day12-part1-simple.py
import time
from typing import List, Dict

class Shape:
    def __init__(self, coords):
        self.coords = coords

def try_place(idx: int, presents: List[int], grid: List[List[int]], 
              width: int, height: int, shapes: Dict[int, Shape]) -> bool:
    if idx >= len(presents):
        return True
    
    shape_idx = presents[idx]
    shape = shapes[shape_idx]
    
    for r in range(height):
        for c in range(width):
            can_place = True
            for dr, dc in shape.coords:
                nr, nc = r + dr, c + dc
                if nr >= height or nc >= width or grid[nr][nc] != -1:
                    can_place = False
                    break
            
            if can_place:
                for dr, dc in shape.coords:
                    grid[r + dr][c + dc] = idx
                
                if try_place(idx + 1, presents, grid, width, height, shapes):
                    return True
                
                for dr, dc in shape.coords:
                    grid[r + dr][c + dc] = -1
    
    return False

def solve():
    with open("../inputs/day-12-input.txt") as f:
        lines = [line.rstrip() for line in f]
    
    shapes = {}
    
    # Parse shapes
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        if not line:
            i += 1
            continue
        
        if line and line[0].isdigit() and ':' in line and 'x' not in line:
            shape_idx = int(line.split(':')[0])
            coords = []
            i += 1
            row = 0
            
            while i < len(lines) and lines[i].strip() and \
                  not (lines[i][0].isdigit() and ':' in lines[i]) and 'x' not in lines[i]:
                for c, ch in enumerate(lines[i]):
                    if ch == '#':
                        coords.append((row, c))
                row += 1
                i += 1
            
            shapes[shape_idx] = Shape(coords)
        else:
            i += 1
    
    # Parse regions
    regions = []
    for line in lines:
        if 'x' in line and ':' in line and line[0].isdigit():
            dims, rest = line.split(':')
            width, height = map(int, dims.split('x'))
            counts = list(map(int, rest.split()))
            
            presents = []
            for idx, count in enumerate(counts):
                presents.extend([idx] * count)
            
            regions.append({'width': width, 'height': height, 'presents': presents})
    
    fittable = 0
    
    for r in regions:
        area = r['width'] * r['height']
        needed = sum(len(shapes[p].coords) for p in r['presents'])
        
        if needed <= area:
            grid = [[-1] * r['width'] for _ in range(r['height'])]
            if try_place(0, r['presents'], grid, r['width'], r['height'], shapes):
                fittable += 1
    
    return fittable

# Benchmark
iterations = 1
start_time = time.perf_counter()

result = 0
for _ in range(iterations):
    result = solve()

end_time = time.perf_counter()
total_time = end_time - start_time
avg_ms = (total_time / iterations) * 1000

print(f"Python (Day 12 Part 1 - Simple)")
print(f"Answer: {result}")
print(f"Average time: {avg_ms:.3f} ms ({iterations} iterations)")
print(f"Total time: {total_time:.2f} s")
